### PR TITLE
Implement a fasttext example

### PR DESCRIPTION
**Pull request by kemaswill**
This commit implements a fasttext example. Fasttext is a very efficient text classification model proposed by A. Joulin, E. Grave, P. Bojanowski and T. Mikolov in the paper "[Bag of Tricks for Efficient Text Classification](https://arxiv.org/abs/1607.01759)". The idea is quite similar with SkipGram, however, instead of predict the current word based on the context words, fasttext predict the category of the document based on the average of all word embeddings in the document.

I have evaluate fasttext model using the IMDB datasets. The fasttext can achieve accuracy of 88.15% after 5 epochs in 70s on a machine with Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, while the CNN achieve accuracy of 88.10% after 1 epochs in 272s.

### CODE REVIEW

**Comment by fchollet** [comment]
> I have evaluate fasttext model using the IMDB datasets. The fasttext can achieve accuracy of 88.15% after 5 epochs in 70s on a machine with Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, while the CNN achieve accuracy of 88.10% after 1 epochs in 272s.

Cool, do you know how that compares to the original FastText and to Vowpal Wabbit, both in speed and accuracy?

**Comment by kemaswill** [comment]
Sure, I will do some experiments to compare this with the Facebook official implementation of fasttext as well as VW.

**Comment by kemaswill** [comment]
I have done some experiments to compare the Facebook official implementation of fasttext v.s. this one. The official fasttext using 20 embedding dimension can achieve accuracy of 82.7%, 86.7%, 87.3%, 87.5% after 5, 10, 15, 20 epochs which take 3s, 7s, 10s, and 13s respectively. So the official one is significant much faster to achieve a similar accuracy.

According to this [post](http://nlpers.blogspot.com/2016/08/fast-easy-baseline-text-categorization.html), it seems that the vw is quite similar with official fasttext on both accuracy and training time.

**Comment by fchollet** [comment]
Consider increasing the batch size to 512. Maybe it would speed things up.

Overall I think it's the `gather` operation in the embedding layer that is slow. Was your benchmark using TF or Theano?

**Comment by kemaswill** [comment]
Increase the batch size from 32 to 512 will improve the per epoch time from 14s to 10.5s. However, the performance is not good and does not always improve after more epochs: 74.97%, 76.02%, 72.69%, 76.47%, 0.7728% for the first 5 epochs.
I am using Theano.

**Comment by kemaswill** [comment]
By the way, I have also tried the TF backend, which gets 88.02% accuracy after 5 epochs in 96s.

File: examples/imdb_fasttext.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+# we add a AveragePooling1D, which
+model.add(AveragePooling1D(pool_length=model.output_shape[1]))
+
+# We flatten the output of the conv layer,
+# so that we can add a dense layer:
+model.add(Flatten())
+
+# We project onto a single unit output layer, and squash it with a sigmoid:
+model.add(Dense(1))
+model.add(Dropout(0.2))
```
Comment (nzw0301) [inline examples/imdb_fasttext.py]: why do you use dropout?

Comment (kemaswill) [inline examples/imdb_fasttext.py]: Just for regularization. Using the dropout will improve the accuracy a little(from 88.13% to 88.15%).

File: examples/imdb_fasttext.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+'''
+
+from __future__ import print_function
+import numpy as np
+np.random.seed(1337)  # for reproducibility
+
+from keras.preprocessing import sequence
+from keras.models import Sequential
+from keras.layers import Dense, Dropout, Activation, Flatten
```
Comment (fchollet) [inline examples/imdb_fasttext.py]: Some of your imports are not used.

File: examples/imdb_fasttext.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+print('Build model...')
+model = Sequential()
+
+# we start off with an efficient embedding layer which maps
+# our vocab indices into embedding_dims dimensions
+model.add(Embedding(max_features,
+                    embedding_dims,
+                    input_length=maxlen))
+
+# we add a AveragePooling1D, which
```
Comment (fchollet) [inline examples/imdb_fasttext.py]: Comment seems truncated

File: examples/imdb_fasttext.py (hunk 1 above)
Comment (fchollet) [inline examples/imdb_fasttext.py]: That is not statistically significant, better remove it (will speed up convergence).

File: examples/imdb_fasttext.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+model.add(AveragePooling1D(pool_length=model.output_shape[1]))
+
+# We flatten the output of the conv layer,
+# so that we can add a dense layer:
+model.add(Flatten())
+
+# We project onto a single unit output layer, and squash it with a sigmoid:
+model.add(Dense(1))
```
Comment (fchollet) [inline examples/imdb_fasttext.py]: Better to include the activation inside this layer rather than as an Activation layer.

**Comment by fchollet** [comment]
LGTM

**Comment by rsteca** [comment]
Hi, I know this issue is closed, but doesn't the original fasttext implementation use char n-grams?

**Comment by AritzBi** [comment]
@rsteca No, according to one of the authors they didn't uset char n-grams for "Bag of tricks for Efficient Text Classification" https://www.facebook.com/groups/1174547215919768/permalink/1192008200840336/?comment_id=1192037870837369&reply_comment_id=1192207620820394&comment_tracking=%7B%22tn%22%3A%22R%22%7D
